downstairs/src/extent.rs

```diff
@@ -0,0 +1,829 @@
+ * a given extent "number".  This directory is generated (see below) when
+ * all the files needed for repairing an extent have been added to the
+ * copy directory and we are ready to start replacing the extents files.
+ * anchored under "dir".  The actual generation of this directory will
+ * be done by renaming the copy directory.
+ */
+pub fn replace_dir<P: AsRef<Path>>(dir: P, number: u32) -> PathBuf {
+    extent_dir(dir, number)
+        .join(extent_file_name(number, ExtentType::Data))
+        .with_extension("replace")
+}
+
+/**
+ * Produce a PathBuf that refers to the completed directory we use for
+ * a given extent "number".  This directory is generated (see below) when
+ * all the files needed for repairing an extent have been copied to their
+ * final destination and everything has been flushed.
+ * The actual generation of this directory will be done by renaming the
+ * replace directory.
+ */
+pub fn completed_dir<P: AsRef<Path>>(dir: P, number: u32) -> PathBuf {
+    extent_dir(dir, number)
+        .join(extent_file_name(number, ExtentType::Data))
+        .with_extension("completed")
+}
+
+/**
+ * Remove directories associated with repair except for the replace
+ * directory. Replace is handled specifically during extent open.
+ */
+pub fn remove_copy_cleanup_dir<P: AsRef<Path>>(dir: P, eid: u32) -> Result<()> {
+    for f in [copy_dir, completed_dir] {
+        let d = f(&dir, eid);
+        if Path::new(&d).exists() {
+            std::fs::remove_dir_all(&d)?;
+        }
+    }
+    Ok(())
+}
+
+impl Extent {
+    fn get_iov_max() -> Result<usize> {
+        let i: i64 = sysconf(SysconfVar::IOV_MAX)?
+            .ok_or_else(|| anyhow!("IOV_MAX returned None!"))?;
+        Ok(i.try_into()?)
+    }
+
+    /**
+     * Open an existing extent file at the location requested.
+     * Read in the metadata from the first block of the file.
+     */
+    pub fn open(
+        dir: &Path,
+        def: &RegionDefinition,
+        number: u32,
+        read_only: bool,
+        log: &Logger,
+    ) -> Result<Extent> {
+        /*
+         * Store extent data in files within a directory hierarchy so that
+         * there are not too many files in any level of that hierarchy.
+         */
+        let path = extent_path(dir, number);
+
+        // If there was an unfinished copy, then clean it up here (??)
+        remove_copy_cleanup_dir(dir, number)?;
+
+        // If the replace directory exists for this extent, then it means
+        // a repair was interrupted before it could finish.  We will continue
+        // the repair before we open the extent.
+        let replace_dir = replace_dir(dir, number);
+        if !read_only && Path::new(&replace_dir).exists() {
+            info!(
+                log,
+                "Extent {} found replacement dir, finishing replacement",
+                number
+            );
+            move_replacement_extent(dir, number as usize, log)?;
+        }
+
```

Comment:
The previous check where we validate file size is.. Ah.
It's moved into the SqliteInner open.
And, that makes sense, as the size check and some of the other checks will
depend on what format the file has, and. yeah, all good.

I should just delete this comment, but I wanted to show you all the hard work
I put into this :)